src/odemis/acq/stream/_sync.py

```diff
@@ -2000,3 +2001,212 @@ def _runAcquisition(self, future):
+        f = model.ProgressiveFuture(start=est_start,
+                                    end=est_start + self.estimateAcquisitionTime())
+        self._current_future = f
+        self._acq_state = RUNNING  # TODO: move to per acquisition
+        self._prog_sum = 0
+        f.task_canceller = self._cancelAcquisition
+
+        # run task in separate thread
+        executeAsyncTask(f, self._runAcquisition, args=(f,))
+        return f
+
+    def _prepareHardware(self):
+        '''
+        Prepare hardware for acquisition and return the best pixel dwelltime value
+        '''
+        # configure the symphotime controller with parameters
+        if hasattr(self._tc_scanner, 'filename') and self._filename is not None:
+            self._tc_scanner.filename.value = self._filename
+        if hasattr(self._tc_scanner, 'directory') and self._directory is not None:
+            self._tc_scanner.directory.value = self._directory
+
+        # The dwell time from the Nikon C2 will set based on what the device is capable of
+        # As a result, we need to recalculate our total dwell time based around this value
+        # and the number of frames we can compute
+        px_dt = min(self._dwellTime.value, self._scanner.dwellTime.range[1])
+        self._nfr = int(math.ceil(self._dwellTime.value / px_dt))  # number of frames
+        px_dt = self._dwellTime.value / self._nfr  # the new dwell time per frame
+        self._scanner.dwellTime.value = px_dt  # try set the C2 dwell time value again.
+        logging.info("Total dwell time: %f s, Pixel Dwell time: %f, collecting %d frames...",
+                     self._dwellTime.value, px_dt, self._nfr)
+
+        self._tc_scanner.dwellTime.value = self._dwellTime.value
+
+        return px_dt
+
+    def _setLightEmission(self, value):
+        # set all light emissions at once to a value
+        em = self._light.emissions.value
+        em = [value] * len(em)
+        self._light.emissions.value = em
+
+    def _runAcquisition(self, future=None):
+
+        logging.debug("Starting job: acquisition")
+
+        try:
+            self._new_acquisition = True
+            self._acq_done.clear()
+
+            px_dt = self._prepareHardware()
+            frame_time = px_dt * numpy.prod(self._scanner.resolution.value)
+            logging.info("Frame time: %f s", frame_time)
+
+            # Start Symphotime acquisition
+            self._time_correlator.data.subscribe(self._onAcqStop)
+
+            # Turn on the lights
+            self._setLightEmission(1)
+
+            # Start the acquisition
+            self._pdetector.data.subscribe(self._onNewData)
+            self._new_acquisition = True
+
+            # For each frame
+            for i in range(self._nfr):
+                # premature cancellation occurred if the flag has already been set.
+                if self._acq_done.is_set():
+                    raise RuntimeError("Acquisition canceled!!!")
+
+                self._frame_done.clear()
```

Comment:
This is quite tricky. If the frame acquisition is fast enough, `frame_done` might have been set() before you arrive here. And you will end up missing it. Maybe what would work better is:
* _onNewData() puts in a queue.Queue the timestamp (or any other stupid object) at which each image has been received
* Here you wait for a new element in the queue, with a timeout.